discovery/xds/xds.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+	"time"
+
+	v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	// Constants for instrumentation.
+	namespace = "prometheus"
+)
+
+// ProtocolVersion is the xDS protocol version
+type ProtocolVersion string
+
+const (
+	ProtocolV3 = ProtocolVersion("v3")
+)
+
+type HTTPConfig struct {
+	config.HTTPClientConfig `yaml:",inline"`
+	RefreshInterval         model.Duration `yaml:"refresh_interval,omitempty"`
+	FetchTimeout            model.Duration `yaml:"fetch_timeout,omitempty"`
+}
+
+// SDConfig is a base config for xDS-based SD mechanisms
+type SDConfig struct {
+	HTTP   HTTPConfig `yaml:",inline"`
+	Server string     `yaml:"server,omitempty"`
+	// ClientID is sent to the xDS management API to identify this client
+	ClientID string `yaml:"client_id,omitempty"`
+}
+
+func init() {
+	// Register top-level SD Configs.
+	discovery.RegisterConfig(&KumaSDConfig{})
+
+	// Register metrics.
+	prometheus.MustRegister(kumaFetchDuration, kumaFetchSkipUpdateCount, kumaFetchFailuresCount)
+
+	// Register protobuf types that need to be marshalled/ unmarshalled.
+	_ = protoTypes.RegisterMessage((&v3.DiscoveryRequest{}).ProtoReflect().Type())
```

Comment:
Do we need the empty assignment? if it is an error, should we deal with it?

Reply:
True, this should be more like a `MustRegister..`